yarn-project/simulator/src/client/logs_cache.ts

```diff
@@ -0,0 +1,67 @@
+import { type EncryptedL2Log, type UnencryptedL2Log } from '@aztec/circuit-types';
+
+/**
+ * Log data that's accessible by all the function calls in an execution.
+ * This class exists to:
+ * 1. Keep track of logs emitted through nested calls in the correct order.
+ * 2. TODO(1641): Remove encrypted logs based on notes nullified in the same scope.
+ */
+export class LogsCache {
+  /**
+   * Logs notes created in this transaction.
+   */
+  private encryptedLogs: EncryptedL2Log[] = [];
+  private unencryptedLogs: UnencryptedL2Log[] = [];
+
+  // TODO Separate encrypted logs linked to note hashes and arbitrary logs:
+
+  // Maps from note hash to encrypted log - useful for removing transient logs
+  // private encryptedLogsLinkedToNotes: Map<bigint, EncryptedL2Log> = new Map();
+
+  // /**
+  //  * Remove the encrypted log for a nullified note.
+  //  * @param innerNoteHash - Inner note hash of the note. If this value equals 0, it means the
+  //  * note being nullified is from a previous transaction (and thus not a new note).
+  //  */
+  // public nullifyNote(innerNoteHash: Fr) {
+  //   if (!innerNoteHash.equals(Fr.ZERO)) {
+  //     // Find and remove the matching new note if the emitted innerNoteHash is not empty.
+  //     const log = this.encryptedLogsLinkedToNotes.get(innerNoteHash.toBigInt()) ?? false;
```

Comment:
There can be more than 1 note that has the same inner note hash. So we might want to associate each log with a note hash counter instead.